src/engine/strat_engine/crypt/shared.rs

```diff
@@ -838,20 +484,55 @@ pub fn activate(
             )));
         }
     }
-    log_on_failure!(
-        device.token_handle().activate_by_token::<()>(
-            Some(&name.to_string()),
-            Some(if unlock_method == UnlockMethod::Keyring {
-                LUKS2_TOKEN_ID
+    if let Some(p) = passphrase {
+        let key_slot =
+            if unlock_method == UnlockMethod::Keyring {
+                Some(get_keyslot_number(device, LUKS2_TOKEN_ID)?.ok_or_else(|| {
+                    StratisError::Msg("LUKS keyring keyslot not found".to_string())
+                })?)
+            } else if unlock_method == UnlockMethod::Clevis {
+                Some(
+                    get_keyslot_number(device, CLEVIS_LUKS_TOKEN_ID)?
+                        .ok_or_else(|| StratisError::Msg("Clevis keyslot not found".to_string()))?,
+                )
             } else {
-                CLEVIS_LUKS_TOKEN_ID
-            }),
-            None,
-            CryptActivate::empty(),
-        ),
-        "Failed to activate device with name {}",
-        name
-    );
+                None
+            };
+        log_on_failure!(
+            device.activate_handle().activate_by_passphrase(
```

Comment:
Supposing you get the correct keyslot, but _also_ supply a bad passphrase, the error message is very low-level: ```Cryptsetup
error: IO error occurred: Operation not permitted (os error 1)```. It might be better to map to something a bit more user-friendly here as well.

Reply:
This error is not specific to this PR and existed previously. I am willing to put up another PR that we can merge separately, but I don't think it belongs here.